### PR TITLE
Feature/Tavily Web Search Driver

### PR DESCRIPTION
**Pull request by william-price01**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
Added TavilyWebSearchDriver
## Issue ticket number and link
Closes https://github.com/griptape-ai/griptape/issues/1167

----
📚 Documentation preview 📚: https://griptape--1179.org.readthedocs.build//1179/

### CODE REVIEW

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+class TavilyWebSearchDriver(BaseWebSearchDriver):
+    api_key: str = field(kw_only=True)
+    client: TavilyClient = field(
+        default=Factory(lambda self: import_optional_dependency("tavily").TavilyClient(self.api_key), takes_self=True),
+        kw_only=True,
+    )
+
+    def search(self, query: str, **kwargs) -> ListArtifact:
+        try:
+            response = self.client.search(query, max_results=self.results_count, **kwargs)
```
[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:26]: Sounds good.

[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:26]: Would you want it added to the BaseWebSearchDriver?

File: docs/griptape-framework/drivers/web-search-drivers.md [hunk 2]
```diff
@@ -9,7 +9,47 @@ Web Search Drivers can be used to search for links from a search query. They are
 
 * `search()` searches the web and returns a [ListArtifact](../../reference/griptape/artifacts/list_artifact.md) that contains JSON-serializable [TextArtifact](../../reference/griptape/artifacts/text_artifact.md)s with the search results.
 
-## Vector Store Drivers
+You can use Web Search Drivers with structures:
```
Comment (collindutter) [inline docs/griptape-framework/drivers/web-search-drivers.md:12]: Capitalize `Structures` and let's turn it into a link.

File: docs/griptape-framework/drivers/web-search-drivers.md [hunk 3]
```diff
@@ -39,3 +73,15 @@ Example of using `DuckDuckGoWebSearchDriver` directly:
 ```python
 --8<-- "docs/griptape-framework/drivers/src/web_search_drivers_3.py"
 ```
+
+### Tavily
+!!! info
+    This driver requires the `drivers-web-search-tavily` [extra](../index.md#extras), and a Tavily [API-KEY](https://app.tavily.com).
```
Comment (collindutter) [inline docs/griptape-framework/drivers/web-search-drivers.md:79]: `API-KEY` -> `api key`

File: docs/griptape-framework/drivers/web-search-drivers.md [hunk 4]
```diff
@@ -39,3 +73,15 @@ Example of using `DuckDuckGoWebSearchDriver` directly:
...
+
+The [TavilyWebSearchDriver](../../reference/griptape/drivers/web_search/tavily_web_search_driver.md) uses the [tavily-python](https://github.com/tavily-ai/tavily-python) SDK for web searching.
```
Comment (collindutter) [inline docs/griptape-framework/drivers/web-search-drivers.md:81]: You can remove this line, no need to mention the underlying SDK.

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/web_search/tavily_web_search_driver.py:26]: This is a similar issue to what I tried solving in https://github.com/griptape-ai/griptape/pull/1003. I think we should solve in a separate PR -- the current solution is consistent with the other Drivers.

File: pyproject.toml
Comment (collindutter) [inline pyproject.toml:1]: Looks like you may have run a toml formatter on this file. While it is a good change, it should not be part of this PR as it makes it difficult to view the diff.

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            return ListArtifact(
+                [
+                    TextArtifact(
+                        json.dumps({"title": result["title"], "url": result["url"], "content": result["content"]})
+                    )
+                    for result in results
+                ]
+            )
+        except Exception as e:
+            raise Exception(f"Error searching '{query}' with Tavily: {e}") from e
```
Comment (collindutter) [inline griptape/drivers/web_search/tavily_web_search_driver.py:37]: I know the other Drivers do it, but catching `Exception` and then immediately re-raising it does not provide much value. @william-price01 are there any more specific exceptions we can catch?

**Review by collindutter — CHANGES_REQUESTED** [review]
Great work! Relatively minor changes.

File: docs/griptape-framework/drivers/web-search-drivers.md [hunk 6]
```diff
@@ -21,12 +61,6 @@ Example using `GoogleWebSearchDriver` directly:
 --8<-- "docs/griptape-framework/drivers/src/web_search_drivers_1.py"
 ```
 
-Example of using `GoogleWebSearchDriver` with an agent:
-
-```python
---8<-- "docs/griptape-framework/drivers/src/web_search_drivers_2.py"
-```
-
```
Comment (vachillo) [inline docs/griptape-framework/drivers/web-search-drivers.md:29]: what happened to this example?

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/drivers/web_search/tavily_web_search_driver.py:26]: i think that makes sense

File: docs/griptape-framework/drivers/web-search-drivers.md (hunk 6 above)
Comment (collindutter) [reply docs/griptape-framework/drivers/web-search-drivers.md:29]: I had instructed @william-price01 to replace "default" example with a DDG one.

**Comment by collindutter** [comment]
@william-price01 after you revert some of the `pyproject.toml` formatting changes, run `poetry lock --no-update` and push that up.

File: pyproject.toml
Comment (william-price01) [reply pyproject.toml:1]: I totally did not mean to, how could I revert without removing the changes?

Comment (collindutter) [reply pyproject.toml:1]: I'd just completely revert and then make your changes again manually

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 5 above)
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:37]: Maybe a missing API-Key? Not really sure. Most of the errors are already handled by the Tavily SDK, they have built in Exception types. Maybe I utilize those?

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/drivers/web_search/tavily_web_search_driver.py:27]: i dont disagree with the fact that most other drivers dont support this but there is a [ton of other params](https://docs.tavily.com/docs/rest-api/api-reference#parameters) on their API that the user now needs to subclass to make use of. the `ProxyWebScraperDriver` does allow passing in arbitrary params. i think we should just follow that model here until we need to change it.

[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:27]: Realizing that this would be helpful in regards to [exa](https://docs.exa.ai/reference/search). They also have a ton of params as well.

File: CHANGELOG.md [hunk 7]
```diff
@@ -26,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Removed `__add__` method from `BaseArtifact`, implemented it where necessary.
 - Generic type support to `ListArtifact`.
 - Iteration support to `ListArtifact`.
+- Moved `language` and `country` from `BaseWebSearchDriver` to `DuckDuckGoWebSearchDriver` and `GoogleWebSearchDriver`
```
Comment (collindutter) [inline CHANGELOG.md:30]: No need to include this since there is no change in functionality

File: pyproject.toml [hunk 8]
```diff
@@ -111,6 +112,7 @@ drivers-web-scraper-trafilatura = ["trafilatura"]
 drivers-web-scraper-markdownify = ["playwright", "beautifulsoup4", "markdownify"]
 
 drivers-web-search-duckduckgo = ["duckduckgo-search"]
+drivers-web-search-tavily = ["tavily-python"]
```
Comment (collindutter) [inline pyproject.toml:115]: Don't forget to add to `all`

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                ]
+            )
+        except MissingAPIKeyError as exc:
+            raise ValueError("API Key is missing, Please provide a valid Tavily API Key.") from exc
+        except UsageLimitExceededError as exc:
+            raise ValueError("Usage Limit Exceeded, Please try again later.") from exc
+        except InvalidAPIKeyError as exc:
+            raise ValueError("Invalid API Key, Please provide a valid Tavily API Key.") from exc
+        except Exception as e:
+            raise ValueError(f"An error occurred while searching for {query} using Tavily: {e}") from e
```
[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:44]: I suppose so, @collindutter what do you think?

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 5 above)
Comment (vachillo) [reply griptape/drivers/web_search/tavily_web_search_driver.py]: not sure that catching specific exceptions just to print and reraise is much better either. should we just remove the `try/except` and let the tool handle it?

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/web_search/tavily_web_search_driver.py:44]: > can we just let the client raise the exception without catching it?

Yeah I think that is better since we're still not doing anything with the caught exceptions.

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:27]: Do we want to add to the base websearch driver? Or just Tavily @collindutter @vachillo

[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/drivers/web_search/tavily_web_search_driver.py:27]: we should just put it into the specific driver for now, and can move it up later if needed

[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py]: Awesome thanks!

File: CHANGELOG.md [hunk 10]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `BaseArtifact.to_bytes()` method to convert an Artifact's value to bytes.
 - `BlobArtifact.base64` property for converting a `BlobArtifact`'s value to a base64 string.
 - `CsvLoader`/`SqlLoader`/`DataframeLoader` `formatter_fn` field for customizing how SQL results are formatted into `TextArtifact`s.
+- `TavilyWebSearchDriver` to utilize Tavily web search SDK.
```
Comment (vachillo) [inline CHANGELOG.md:15]: move this into the top `Unreleased` section under a new `Added` header

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+
+@define
+class TavilyWebSearchDriver(BaseWebSearchDriver):
+    api_key: str = field(kw_only=True)
+    client: TavilyClient = field(
+        default=Factory(lambda self: import_optional_dependency("tavily").TavilyClient(self.api_key), takes_self=True),
+        kw_only=True,
+    )
+    params: dict = field(default=Factory(dict), kw_only=True, metadata={"serializable": True})
```
Comment (vachillo) [inline griptape/drivers/web_search/tavily_web_search_driver.py:23]: lets update this to:
```suggestion
    params: dict[str, Any] = field(factory=dict, kw_only=True, metadata={"serializable": True})
```

File: pyproject.toml [hunk 12]
```diff
@@ -71,6 +71,7 @@ pypdf = {version = "^3.9", optional = true}
 pillow = {version = "^10.2.0", optional = true}
 mail-parser = {version = "^3.15.0", optional = true}
 filetype = {version = "^1.2", optional = true}
+tavily-python = "^0.5.0"
```
Comment (vachillo) [inline pyproject.toml:74]: lets move this under `torch` and follow the convention of the other imports by adding `optional = true`

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 11 above)
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:23]: Sounds good!

File: CHANGELOG.md (hunk 10 above)
Comment (william-price01) [reply CHANGELOG.md:15]: Sweet

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        kw_only=True,
+    )
+    params: dict[str, Any] = field(factory=dict, kw_only=True, metadata={"serializable": True})
+
+    def search(self, query: str, **kwargs) -> ListArtifact:
+        response = self.client.search(query, max_results=self.results_count, **self.params, **kwargs)
+        results = response["results"]
+        return ListArtifact(
+            [
+                TextArtifact(json.dumps({"title": result["title"], "url": result["url"], "content": result["content"]}))
```
Comment (vachillo) [inline griptape/drivers/web_search/tavily_web_search_driver.py:30]: @collindutter would this be a good use of `JsonArtifact`?

File: pyproject.toml (hunk 12 above)
Comment (william-price01) [reply pyproject.toml]: missed that, thanks

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 13 above)
Comment (collindutter) [reply griptape/drivers/web_search/tavily_web_search_driver.py:30]: Let's give it a go

File: CHANGELOG.md [hunk 14]
```diff
@@ -5,7 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
-
+### Added
+- `TavilyWebSearchDriver` to utilize Tavily web search SDK.
```
Comment (collindutter) [inline CHANGELOG.md:9]: Using the web search SDK is kind of a low level detail. Try rephrasing to emphasize that we're using some feature of Tavily's service.

File: griptape/drivers/web_search/tavily_web_search_driver.py (hunk 13 above)
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:30]: Alright sounds good!

File: CHANGELOG.md [hunk 15]
```diff
@@ -5,7 +5,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
-
+### Added
+- Introduced `TavilyWebSearchDriver` to integrate Tavily's web search capabilities:
+  - Supports configurable search parameters
+  - Returns search results as structured artifacts
```
Comment (collindutter) [inline CHANGELOG.md:11]: You can drop this part.

File: CHANGELOG.md [hunk 16]
```diff
@@ -5,7 +5,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
-
+### Added
+- Introduced `TavilyWebSearchDriver` to integrate Tavily's web search capabilities:
```
Comment (collindutter) [inline CHANGELOG.md:9]: `Introduced` can be dropped. The `Added` and `Fixed` are implied when reading.

File: CHANGELOG.md [hunk 17]
```diff
@@ -20,6 +23,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Iteration support to `ListArtifact`.
 
 
+
```
Comment (collindutter) [inline CHANGELOG.md:26]: Nit: try to avoid introducing unnecessary whitespace changes

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    api_key: str = field(kw_only=True)
+    client: TavilyClient = field(
+        default=Factory(lambda self: import_optional_dependency("tavily").TavilyClient(self.api_key), takes_self=True),
+        kw_only=True,
+    )
+    params: dict[str, Any] = field(factory=dict, kw_only=True, metadata={"serializable": True})
+
+    def search(self, query: str, **kwargs) -> ListArtifact:
+        response = self.client.search(query, max_results=self.results_count, **self.params, **kwargs)
+        results = response["results"]
```
Comment (collindutter) [inline griptape/drivers/web_search/tavily_web_search_driver.py:26]: Is `results` always populated? What happens in an exception scenario?

File: pyproject.toml [hunk 19]
```diff
@@ -72,6 +73,7 @@ pillow = {version = "^10.2.0", optional = true}
 mail-parser = {version = "^3.15.0", optional = true}
 filetype = {version = "^1.2", optional = true}
 
+
```
Comment (collindutter) [inline pyproject.toml:76]: Nit: try to avoid introducing unnecessary whitespace changes

**Review by collindutter — CHANGES_REQUESTED** [review]
Nice work, just a couple minor comments.

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+        return ListArtifact([TextArtifact(JsonArtifact(result)) for result in results])
```
Comment (vachillo) [inline griptape/drivers/web_search/tavily_web_search_driver.py:27]: shouldnt be re-wrapped in a text artifact.
```suggestion
        return ListArtifact([JsonArtifact(result) for result in results])
```
is something downstream failing?

File: griptape/drivers/web_search/tavily_web_search_driver.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+    @lazy_property()
+    def client(self) -> TavilyClient:
+        return import_optional_dependency("tavily").TavilyClient(self.api_key)
+
+    def search(self, query: str, **kwargs) -> ListArtifact:
+        response = self.client.search(query, max_results=self.results_count, **self.params, **kwargs)
+        results = response.get("results", [])
+        if not results:
+            raise BadRequestError("No results found or the response structure is invalid.")
```
[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:31]: Was adjusting based on a review that Collin gave, he was asking what happens when results is empty. 

[Discussion](https://github.com/griptape-ai/griptape/pull/1179#discussion_r1771781170)

[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/web_search/tavily_web_search_driver.py:31]: Poor phrasing on my end -- I was asking if `results` is always present in the dictionary. Empty results are fine.

[the remark this answers is not in the document]
Comment (william-price01) [reply griptape/drivers/web_search/tavily_web_search_driver.py:31]: That's how the api-spec is defined. So unless they change the schema anytime soon, should be fine.

File: CHANGELOG.md [hunk 22]
```diff
@@ -8,7 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Added
 - Parameter `pipeline_task` on `HuggingFacePipelinePromptDriver` for creating different types of `Pipeline`s.
-
+- `TavilyWebSearchDriver` to integrate Tavily's web search capabilities.
```
Comment (vachillo) [inline CHANGELOG.md:11]: nit, add newline

File: CHANGELOG.md [hunk 23]
```diff
@@ -30,8 +30,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - `model_client` on `GoogleTokenizer`.
 - **BREAKING**: Renamed parameter `pipe` on `HuggingFacePipelinePromptDriver` to `pipeline`.
 - Several places where API clients are initialized are now lazy loaded.
-
-
```
Comment (vachillo) [inline CHANGELOG.md:34]: nit: add these back

**Review by vachillo — COMMENTED** [review]
looks good once the other small whitespace changes are addressed

File: CHANGELOG.md (hunk 22 above)
Comment (collindutter) [reply CHANGELOG.md:12]: And remove newline on line 11

**Review by collindutter — APPROVED** [review]
Great work and congrats on the first PR!
